commit ebc65295f5ab943955ea6ae9772932c32e39d02c
Author: Christopher Faulet <cfaulet@haproxy.com>
Date:   Wed Apr 10 14:21:51 2019 +0200

    BUG/MEDIUM: spoe: Return an error if nothing is encoded for fragmented messages
    
    If the maximum frame size is very small with a large message or argument name,
    it is possible to be unable to encode anything. In such case, it is important to
    stop processing returning an error otherwise we will retry in loop to encode the
    message, failing each time because of the too small frame size.
    
    This patch must be backported to 1.9 and 1.8.
    
    (cherry picked from commit a715ea82eacf4ccf7f447bf4dd4111cc29fe171e)
    Signed-off-by: William Lallemand <wlallemand@haproxy.org>
    (cherry picked from commit 3c76e4d79669329ae972f3348e441fea7316813f)
    [cf: Adapted to use old buffer API]
    Signed-off-by: Christopher Faulet <cfaulet@haproxy.com>

diff --git a/src/flt_spoe.c b/src/flt_spoe.c
index 64601e3f..95f30898 100644
--- a/src/flt_spoe.c
+++ b/src/flt_spoe.c
@@ -2276,7 +2276,9 @@ spoe_encode_messages(struct stream *s, struct spoe_context *ctx,
 	return 1;
 
   too_big:
-	if (!(agent->flags & SPOE_FL_SND_FRAGMENTATION)) {
+	/* Return an error if fragmentation is unsupported or if nothing has
+	 * been encoded because its too big and not splittable. */
+	if (!(agent->flags & SPOE_FL_SND_FRAGMENTATION) || p == ctx->buffer->p) {
 		ctx->status_code = SPOE_CTX_ERR_TOO_BIG;
 		return -1;
 	}
